src/cryptography/hazmat/primitives/serialization/ssh.py

```diff
@@ -4,150 +4,679 @@
-        )
+def _check_empty(data):
+    """All data should have been parsed.
+    """
+    if data:
+        raise ValueError("Corrupt data: unparsed data")
 
-    return loader(key_type, rest, backend)
 
+def _init_cipher(ciphername, password, salt, rounds, backend):
+    """Generate key + iv and return cipher.
+    """
+    if not password:
+        raise ValueError("Key is password-protected.")
 
-def _load_ssh_rsa_public_key(key_type, decoded_data, backend):
-    e, rest = _ssh_read_next_mpint(decoded_data)
-    n, rest = _ssh_read_next_mpint(rest)
+    algo, key_len, mode, iv_len = _SSH_CIPHERS[ciphername]
+    seed = _bcrypt_kdf(password, salt, key_len + iv_len, rounds, True)
+    return Cipher(algo(seed[:key_len]), mode(seed[key_len:]), backend)
 
-    if rest:
-        raise ValueError('Key body contains extra bytes.')
 
-    return rsa.RSAPublicNumbers(e, n).public_key(backend)
+def _get_u32(data):
+    """Uint32"""
+    if len(data) < 4:
+        raise ValueError("Invalid data")
+    return _U32.unpack(data[:4])[0], data[4:]
 
 
-def _load_ssh_dss_public_key(key_type, decoded_data, backend):
-    p, rest = _ssh_read_next_mpint(decoded_data)
-    q, rest = _ssh_read_next_mpint(rest)
-    g, rest = _ssh_read_next_mpint(rest)
-    y, rest = _ssh_read_next_mpint(rest)
+def _get_u64(data):
+    """Uint64"""
+    if len(data) < 8:
+        raise ValueError("Invalid data")
+    return _U64.unpack(data[:8])[0], data[8:]
 
-    if rest:
-        raise ValueError('Key body contains extra bytes.')
 
-    parameter_numbers = dsa.DSAParameterNumbers(p, q, g)
-    public_numbers = dsa.DSAPublicNumbers(y, parameter_numbers)
+def _get_sshstr(data):
+    """Bytes with u32 length prefix"""
+    n, data = _get_u32(data)
+    if n > len(data):
+        raise ValueError("Invalid data")
+    return data[:n], data[n:]
 
-    return public_numbers.public_key(backend)
 
+def _get_mpint(data):
+    """Big integer."""
+    val, data = _get_sshstr(data)
+    if val and six.indexbytes(val, 0) > 0x7F:
+        raise ValueError("Invalid data")
+    return utils.int_from_bytes(val, 'big'), data
 
-def _load_ssh_ecdsa_public_key(expected_key_type, decoded_data, backend):
-    curve_name, rest = _ssh_read_next_string(decoded_data)
-    data, rest = _ssh_read_next_string(rest)
 
-    if expected_key_type != b"ecdsa-sha2-" + curve_name:
-        raise ValueError(
-            'Key header and key body contain different key type values.'
-        )
+def _to_mpint(val):
+    """Storage format for signed bigint.
+    """
+    if val < 0:
+        raise ValueError("negative mpint not allowed")
+    if not val:
+        return b""
+    nbytes = (val.bit_length() + 8) // 8
+    return utils.int_to_bytes(val, nbytes)
 
-    if rest:
-        raise ValueError('Key body contains extra bytes.')
 
-    curve = {
-        b"nistp256": ec.SECP256R1,
-        b"nistp384": ec.SECP384R1,
-        b"nistp521": ec.SECP521R1,
-    }[curve_name]()
+class _FragList:
```

Comment:
This will generate an old-style class in Python 2.7. Since nobody should subclass this or do any other nonsense it probably doesn't matter, but let's just have it inherit from `object` to be consistent with all our other class declarations until the wonderful day we can drop 2.7 support. (Please also do this for the other classes below this)

Reply:
Fixed